### ISSUE TITLE
MongoDB: `upsert()` fails on `update` with relation when `notablescan: 1`

### ISSUE THREAD

**Issue by paulrostorp** [prisma/prisma#12793]
### Bug description

Using MongoDB, the following error occurs when running an upsert and the document already exists:

```
PrismaClientUnknownRequestError:
Invalid `prisma.user.upsert()` invocation:


  Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: RawDatabaseError { code: "unknown", message: "Command failed (NoQueryExecutionPlans): error processing query: ns=ninja-diary.projectsTree: $expr {$and: [{$in: [\"$ownerId\", {$const: [\"qGdWa0VNzHfW8GRaDxz5nfjPDRt1\"]}]}, {$or: [{$ne: [{$ifNull: [\"$ownerId\", {$const: null}]}, {$const: null}]}, {$eq: [\"$ownerId\", {$const: null}]}]}]}Sort: {}\nProj: { _id: true, ownerId: true }\nCollation: { locale: \"simple\" }\n planner returned error :: caused by :: No indexed plans available, and running with 'notablescan')" } })
    at Object.request (/usr/app/node_modules/@prisma/client/runtime/index.js:45582:15)
    at async PrismaClient._request (/usr/app/node_modules/@prisma/client/runtime/index.js:46405:18)
    at async setSyncTimeForUser (/usr/app/src/features/sync/utils.ts:29:18)
    at async /usr/app/src/features/sync/push.ts:19:5
    at async Proxy._transactionWithCallback (/usr/app/node_modules/@prisma/client/runtime/index.js:46366:18)
    at async push (/usr/app/src/features/sync/push.ts:15:3)
    at async /usr/app/src/app.ts:442:7 {
  clientVersion: '3.12.0'
}
```

### How to reproduce

```prisma
model User {
  id       String    @id @map("_id") @db.String
  name String
  lastSync DateTime  @updatedAt
  projects Project[]

  @@map("users")
}

model Project {
  id        String   @id @map("_id") @db.String
  name      String
  owner     User     @relation(fields: [ownerId], references: [id])
  ownerId   String   @db.String
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  isDeleted Boolean  @default(false)
  
  @@index([ownerId])
  @@index([createdAt])
  @@index([updatedAt])
  @@index([isDeleted])
  @@map("projects")
}
```
Run this once:
```javascript
  await prisma.user.upsert({
    where: { id: userId },
    create: { id: userId, name: "foo" },
    update: { name: "foo2" },
  });
```

This will create the document, which works.
Then run it a second time, which will execute the "update" part since the document exists. It now fails with the error above.

If I remove the relation from the schema (and only have a "dum" `ownerId` field) the opertation works again.

### Expected behavior

It should not fail

### Prisma information

Prisma v3.12.0

mongoDB 4.4

```prisma
model User {
  id       String    @id @map("_id") @db.String
  name String
  lastSync DateTime  @updatedAt
  projects Project[]

  @@map("users")
}

model Project {
  id        String   @id @map("_id") @db.String
  name      String
  owner     User     @relation(fields: [ownerId], references: [id])
  ownerId   String   @db.String
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  isDeleted Boolean  @default(false)
  
  @@index([ownerId])
  @@index([createdAt])
  @@index([updatedAt])
  @@index([isDeleted])
  @@map("projects")
}
```
Run this once:
```javascript
  await prisma.user.upsert({
    where: { id: userId },
    create: { id: userId, name: "foo" },
    update: { name: "foo2" },
  });
```

### Environment & setup

- OS: debian
- Database: MongoDB Atlas v4.4
- Node.js version: 16.13

### Prisma Version

```
prisma                  : 3.12.0
@prisma/client          : 3.12.0
Current platform        : debian-openssl-1.1.x
Query Engine (Node-API) : libquery-engine 22b822189f46ef0dc5c5b503368d1bee01213980 (at node_modules/@prisma/engines/libquery_engine-debian-openssl-1.1.x.so.node)
Migration Engine        : migration-engine-cli 22b822189f46ef0dc5c5b503368d1bee01213980 (at node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Introspection Engine    : introspection-core 22b822189f46ef0dc5c5b503368d1bee01213980 (at node_modules/@prisma/engines/introspection-engine-debian-openssl-1.1.x)
Format Binary           : prisma-fmt 22b822189f46ef0dc5c5b503368d1bee01213980 (at node_modules/@prisma/engines/prisma-fmt-debian-openssl-1.1.x)
Default Engines Hash    : 22b822189f46ef0dc5c5b503368d1bee01213980
Studio                  : 0.459.0
Preview Features        : interactiveTransactions, extendedIndexes
```

**Comment by janpio** [comment]
Hm, I can not reproduce the problem here.

I can run this script many times with the schema you provided:
```
const { PrismaClient } = require('@prisma/client')

const prisma = new PrismaClient()

async function main() {
  let userId = "3"
  await prisma.user.upsert({
    where: { id: userId },
    create: { id: userId, name: "foo" },
    update: { name: "foo2" },
  });
}

main()
  .catch((e) => {
    console.error(e)
    process.exit(1)
  })
  .finally(async () => {
    await prisma.$disconnect()
  })
```
This is tested with MongoDB Atlas Serverless, which runs 5.3.1.

(Waiting for 4.4 to be provisioned now)

**Comment by janpio** [comment]
I can "unfortunately" also not reproduce this with a MongoDB Atlas running 4.4.

Did you try the reproduction you provided in a new database? Maybe the existing data or stat of the server influences what you are seeing.

**Comment by dpetrick** [comment]
I am also unable to reproduce the issue.

**Comment by paulrostorp** [comment]
@janpio @dpetrick Was `notablescan` enabled on the cluster you tried ? (`Require Indexes for All Queries` option in advanced cluster configuration)

**Comment by paulrostorp** [comment]
It seems the operation fails at this step: 
```
db.Proj.aggregate([ { $match: { $expr: { $and: [ { }, { $and: [ { $in: [ "$ownerId", [ "x", ], ], }, { $or: [ { $ne: [ { $ifNull: [ "$ownerId", null, ], }, null, ], }, { $eq: [ "$ownerId", null, ], }, ], }, ], }, ], }, }, }, { $project: { _id: 1, ownerId: 1, }, }, ])
```
But I cannot figure out which index is missing...

**Comment by janpio** [comment]
<img width="310" alt="image" src="https://user-images.githubusercontent.com/183673/166591326-8ebd2a8e-5d8a-436f-9b9e-cafed24933d2.png">

**Comment by paulrostorp** [comment]
@janpio yes please 😄

You can also enable with mongo client / shell 
```
db.admin().command({ setParameter: 1, notablescan: 1 });
```

Sorry for not specifying this earlier, I was unaware of the existence of this configuration setting until my followup investigation yesterday.

**Comment by paulrostorp** [comment]
I've found a lot more cases where this happens. It seems the generated `aggregate` commands fails to find an index when performing operations only on the `_id` field. 

For example, this fails:
```js
txn.project.updateMany({
  where: {
    id: { in: projectIds }, // array of project ids
  },
  data: {
    updatedAt: timestamp,
    isDeleted: true,
  },
});
```

But the following succeeds:

```js
txn.project.updateMany({
  where: {
    id: { in: projectIds }, // array of project ids
    isDeleted: false
  },
  data: {
    updatedAt: timestamp,
    isDeleted: true,
  },
});
```

**Comment by janpio** [comment]
At least the warning is explicit: 
<img width="299" alt="image" src="https://user-images.githubusercontent.com/183673/166660364-dc095fb4-f775-4885-b57b-adaf85a52b8d.png">
Will test when configuration is applied.

(If you have the time and means, figuring out if this also happens with newer versions would be interesting as well)

**Comment by paulrostorp** [comment]
@janpio I am able to reproduce locally with mongo v5.0.6 and prisma `v3.13.0`

**Comment by paulrostorp** [comment]
I did some more investigating and the issue seems broader, as a lot of operations seem to cause a collection scan instead of using an index.

Specifically, I have found that using the `in` operator always causes an aggregation command that fails to use an index. For example
```js
prisma.project.findMany({
  where: {
    id: { in: [uuid()] },
  },
});
```
causes a collection scan. I have confirmed this by running `explain` on the `aggregate` command prisma ends up running (using query logging).

I added a couple million objects to the collection, and running the `findMany` command takes over 300ms locally vs <5ms when using mongoDB driver

**Comment by janpio** [comment]
Temporary conclusion then if we can reproduce any of this: Prisma MongoDB is currently not compatible with `notablescan: 1` - correct?

**Comment by paulrostorp** [comment]
@janpio yes correct.

**Comment by janpio** [comment]
I can unfortunately confirm this:

```
PS C:\Users\Jan\Documents\throwaway\12793> node .\script.js
PrismaClientUnknownRequestError: 
Invalid `prisma.user.upsert()` invocation:


  Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: RawDatabaseError { code: "unknown", message: "Command failed (NoQueryExecutionPlans): error processing query: ns=12793.projectsTree: $expr {$and: [{$in: [\"$ownerId\", {$const: [\"3\"]}]}, {$or: [{$ne: [{$ifNull: [\"$ownerId\", {$const: null}]}, {$const: null}]}, {$eq: [\"$ownerId\", {$const: null}]}]}]}Sort: {}\nProj: { _id: true, ownerId: true }\nCollation: { locale: \"simple\" }\n planner returned error :: caused by :: No indexed plans available, and running with 'notablescan')" } }) 
    at Object.request (C:\Users\Jan\Documents\throwaway\12793\node_modules\@prisma\client\runtime\index.js:45633:15)
    at async PrismaClient._request (C:\Users\Jan\Documents\throwaway\12793\node_modules\@prisma\client\runtime\index.js:46456:18)
    at async main (C:\Users\Jan\Documents\throwaway\12793\script.js:8:5) {
  clientVersion: '3.13.0'
}
```

(Tested with MongoDB 4.4 on MongoDB Atlas, via setting `Configuration => Additional Setings => More Configuration Options => Require Indexes for All Queries => "On"`)

Now that we know that, would you expect this query to work (= necessary index being present) with your current schema?

**Comment by janpio** [comment]
Simplified schema for further reproduction:

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

model User {
  id       String    @id @map("_id") @db.String
  name String
  projects Project[]
}

model Project {
  id        String   @id @map("_id") @db.String
  owner     User     @relation(fields: [ownerId], references: [id])
  ownerId   String   @db.String
}
```

and script with query logging:
```js
const { PrismaClient } = require('@prisma/client')

const prisma = new PrismaClient({ log: ["query"]})

// A `main` function so that we can use async/await
async function main() {
    let userId = "3"
    await prisma.user.upsert({
        where: { id: userId },
        create: { id: userId, name: "foo" },
        update: { name: "foo2" },
      });
}

main()
  .catch((e) => {
    console.error(e)
    process.exit(1)
  })
  .finally(async () => {
    await prisma.$disconnect()
  })
```

**Comment by janpio** [comment]
Internal discussion about the query and index usage: https://prisma-company.slack.com/archives/C0239SYUZV1/p1652120382980849
TLDR: The current query structure makes it really hard / impossible for the query to be supported by indexes, because of the `$or` (which makes sure it's not null although we know we are searching for a value).

**Comment by paulrostorp** [comment]
@janpio To answer your question, I do not think it is a missing index on my end. I have spent a very long time investigating this and tested all kinds of indexes on all collections.

FYI: Much simpler operations also fail (without relations)
```prisma
```
```js
prisma.project.findMany({
  where: {
    id: { in: [uuid()] },
  },
});
```

**Comment by janpio** [comment]
As written, the way the query above is generated internally by Prisma it is really hard / impossible for an index to be used.

The same might be true for the other one you just posted. Optimally enable query logging and post the `prisma:query` output for that Prisma Client query as well.

**Comment by paulrostorp** [comment]
I've created a benchmark against mongoose, it is described in more details here: https://github.com/prisma/prisma/discussions/14811

**Comment by paulrostorp** [comment]
Is there an ETA on when the performance issues outlined in this issue and https://github.com/prisma/prisma/issues/14871 will be addressed ?

I was hoping for some improvement before I had to go to production, but that is looking less and less likely, so I am planning to replace prisma with mongoose in all our mongo-backed services.

Is there any chance for performance improvement within the next 1-2 months ?

**Comment by AndreasKarz** [comment]
The 'problem' is in your client config - search for

mongoRsResource.**Client.DisableTableScan()**;

This prevents to query to fields without index. **If you enable TableScan, you can shoot your performance into nirvana if you have a large database.**

**Comment by paulrostorp** [comment]
@AndreasKarz the issue is that the prisma client generates overly complex queries, which force a table scan. My db has disabled table scans.

**Comment by peymil** [comment]
> As written, the way the query above is generated internally by Prisma it is really hard / impossible for an index to be used.
> 
> The same might be true for the other one you just posted. Optimally enable query logging and post the `prisma:query` output for that Prisma Client query as well.

Is performance problem mentioned in Prisma documentation? Because this is a huge factor while choosing an ORM.

**Comment by domharrington** [comment]
> This prevents to query to fields without index. **If you enable TableScan, you can shoot your performance into nirvana if you have a large database.**

@AndreasKarz the intention isn't to run with `notablescan` in production, but to run with it in development to catch potential performance issues in development and add appropriate indexes before moving to production to ensure the fast running of queries.

**Comment by AmineHouari98** [comment]
any updates on this issue ? seems like indexes are not used either with `include` parameter.

**Comment by paulrostorp** [comment]
@AmineHouari98 I encourage you to add your case to https://github.com/prisma/prisma/issues/14871 there are many queries with performance issues that are not documented

**Comment by DoubleJ-G** [comment]
I am also having issues where trying to just find documents greater than a certain date

```ts
const start = dayjs(Date.now()).subtract(8, 'hours').toDate();

return this.prismaService.tarp.findMany({
  where: {
    timestamp: {
       gt: start,
     },
  },
});
```

Ends up in a query like 

```
"db.TARPDocuments.aggregate([ { $match: { $expr: { $and: [ { $gt: [ \"$timestamp\", { $literal: DateTime(\"2023-06-02 6:59:24.852 +00:00:00\"), }, ], }, { $ne: [ \"$timestamp\", \"$$REMOVE\", ], }, ], }, }, }, { $project: { _id: 1, installationId: 1, vehicleId: 1, timestamp: 1, tyreId: 1, hidden: 1, overridden: 1, injuryTemperature: 1, injuryHeight: 1, injuryWidth: 1, tyreAbsoluteTemp: 1, tyreDifferentialTemp: 1, injuryDifferentialTemp: 1, }, }, ])"
```

Due to the $expr this doesn't seem to be taking advantage of indexes on the database resulting in a query time of 2s as it tries to search through hundreds of thousands of documents.

**Comment by yufangOPUS** [comment]
> any updates on this issue ? seems like indexes are not used either with `include` parameter.

Same here, just found our atlas have so much collscan for in queries but the fields are all created unique index on

**Comment by jvgaeta** [comment]
> any updates on this issue ? seems like indexes are not used either with `include` parameter.

We're encountering the same issue with 'include' after going to production. We were able to hack around it for now. However, we will likely migrate off of Prisma if a fix doesn't come sometime soon which stinks since the DX has been amazing outside of the performance issues w/ mongodb.

**Comment by off99555** [comment]
> This (using `in`) has just caused us an enormous ($$$) bill on mongodb atlas (serverless bills by usage), due to prisma running full table scans and ignoring indexes. I was shocked and disappointed to learn that this is known issue without a workaround. There should be a warning for everyone to _not_ use prisma with mongodb currently.

Yep, if I know before hand that **bad performance** means not using index and costing me 100x more then I wouldn't have gone with Prisma. I would have used something like Mongoose instead.

**Comment by boredland** [comment]
> something like Mongoose instead

My recommendation is https://mikro-orm.io/. Haven't used it with MongoDB, but with PG it was amazing. And of course migrating off MongoDB mid-term in most use-cases.

**Comment by shaikegross** [comment]
This is extremely disappointing and a bit deceiving to hide from Prisma Mongo users. the fact a simple where _id in [] won't use an index is shocking. We will be migrating off Prisma. and to think we were considering accelerate 😞

**Comment by janpio** [comment]
Speaking of, the original problem with `upsert()` when it goes into `update` has also been fixed in 5.14.0.

Here the original reproduction with 3.12.0 again:
```
PS C:\Users\Jan\Documents\throwaway\12793> node .\script.js  
prisma:query db.User.aggregate([ { $match: { $expr: { $and: [ { $and: [ { $eq: [ "$_id", "4", ], }, { $or: [ { $ne: [ { $ifNull: [ "$_id", null, ], }, null, ], }, { $eq: [ "$_id", null, ], }, ], }, ], }, ], }, }, }, { $project: { _id: 1, }, }, ])
prisma:query db.Project.aggregate([ { $match: { $expr: { $and: [ { }, { $and: [ { $in: [ "$ownerId", [ "4", ], ], }, { $or: [ { $ne: [ { $ifNull: [ "$ownerId", null, ], }, null, ], }, { $eq: [ "$ownerId", null, ], }, ], }, ], }, ], }, }, }, { $project: { _id: 1, ownerId: 1, }, }, ])
PrismaClientUnknownRequestError: 
Invalid `prisma.user.upsert()` invocation:


  Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: RawDatabaseError { code: "unknown", message: "Command failed (NoQueryExecutionPlans): error processing query: ns=12793repro2.ProjectTree: $expr {$and: [{$in: [\"$ownerId\", {$const: [\"4\"]}]}, {$or: [{$ne: [{$ifNull: [\"$ownerId\", {$const: null}]}, {$const: null}]}, {$eq: [\"$ownerId\", {$const: null}]}]}]}\nSort: {}\nProj: { _id: true, ownerId: true }\nCollation: { locale: \"simple\" }\n planner returned error :: caused by :: No indexed plans available, and running with 'notablescan')" } })
    at Object.request (C:\Users\Jan\Documents\throwaway\12793\node_modules\@prisma\client\runtime\index.js:45582:15)
    at async PrismaClient._request (C:\Users\Jan\Documents\throwaway\12793\node_modules\@prisma\client\runtime\index.js:46405:18)
    at async main (C:\Users\Jan\Documents\throwaway\12793\script.js:8:5) {
  clientVersion: '3.12.0'
}
```

And the same script run with 5.14.0:
```
PS C:\Users\Jan\Documents\throwaway\12793> node .\script.js
prisma:query db.User.aggregate([ { $match: { $expr: { $and: [ { $and: [ { $and: [ { $eq: [ "$_id", { $literal: "4", }, ], }, { $ne: [ "$_id", "$$REMOVE", ], }, ], }, ], }, { }, ], }, }, }, { $project: { _id: 1, }, }, ])
prisma:query db.User.updateMany({ _id: { $in: [ "4", ], }, },{ $set: { name: { $literal: "foo2", }, }, })
prisma:query db.User.aggregate([ { $match: { $expr: { $and: [ { $and: [ { $eq: [ "$_id", { $literal: "4", }, ], }, { $ne: [ "$_id", "$$REMOVE", ], }, ], }, ], }, }, }, { $project: { _id: 1, name: 1, }, }, ])
```

Yay!
